index.rst

```diff
@@ -1,29 +1,413 @@
+This necessitates that a vertical monorepo must have two directories at its root to host separate Python projects for the client and server:
+
+.. code-block:: text
+   :name: layout
+   :caption: Vertical client-server monorepo layout (abridged)
+
+   example
+   ├── .github
+   │   ├── dependabot.yml
+   │   └── workflows
+   ├── .pre-commit-config.yaml
+   ├── client
+   │   ├── pyproject.toml
+   │   └── src
+   │       └── exampleclient
+   │           ├── __init__.py
+   │           └── models.py
+   ├── Dockerfile
+   ├── Makefile
+   └── server
+       ├── pyproject.toml
+       ├── requirements
+       └── src
+           ├── example
+           │   ├── __init__.py
+           │   ├── config.py
+           │   ├── dependencies
+           │   ├── domain
+           │   ├── main.py
+           │   ├── handlers
+           │   └── services
+           └── tests
+
+This monorepo contains two Python packages: ``example`` (the application) and ``exampleclient`` (the library).
+The ``exampleclient.models`` module contains the Pydantic classes that define the REST API for the ``example`` application.
+
+How the application depends on the client library
+-------------------------------------------------
+
+For an effective development workflow, the application needs to be able to import models from the client library locally, rather than through a PyPI release.
+In current practice, applications use the ``requirements.txt`` file format to declare their dependencies.
+We were not able to declare a local dependency in the requirements file, though.
+
+We found the only viable mechanism is to manually pip install the client library in development and deployment contexts (the specific patterns are explored below).
+The downside of this approach is that the client isn't considered by the pinned dependencies compiled by pip-tools_.
+Normally runtime dependencies for the server application are abstractly listed in a ``requirements/main.in`` file for each application; pip-tools_ compiles these dependencies and their sub-dependencies into a ``requirements/main.txt`` file which is committed to the Git repository and actually used for installing dependencies.
+This practice ensures that Docker builds and development environments alike are reproducible.
+In practice, the client library's absence from ``requirements/main.txt`` itself isn't harmful because the client is inherently pinned by virtue of being co-developed in the same Git repository.
+
+What's potentially concerning, though, is the absence of the client's own dependencies from the application's ``requirements/main.txt`` dependencies.
+We could mitigate this risk by limiting client library dependencies to packages that are already in the main application's ``requirements/main.txt``.
```

Comment:
Maybe something to the effect of "we don't expect this to be onerous to developers because client models are generally subsets of richer server models."